keps/sig-etcd/4578-feature-gate/README.md

```diff
@@ -0,0 +1,388 @@
+| Beta | <ul><li>Enabled by default. </li><li>The feature is well tested. Enabling the feature is considered safe.</li><li>Support for the overall feature will not be dropped, though details may change.</li><li>Recommended for only non-business-critical uses because of potential for discovering new hard-to-spot bugs through wider adoption.</li></ul> | Before moving a feature to GA, it should have <ul><li> Widespread usage.</li><li>No bug reported for at least 1 minor release.</li><li>Other feature specific criteria in the feature KEP. </li></ul> |
+| GA | <ul><li>The feature is always enabled; you cannot disable it.</li><li>The corresponding feature gate is no longer needed.</li><li>Stable versions of features will appear in released software for many subsequent versions.</li></ul> | Before deprecating a GA feature, it should have <ul><li> Feature deprecation announcement.</li><li>No user impacting change for at least 1 minor release.</li><li>Other feature specific criteria in the feature KEP. </li></ul> |
+| Deprecated | <ul><li>The feature gate is no longer in use. </li><li>The feature has graduated to GA or been removed.</li></ul> | <ul><li>If deprecating Beta feature, should set default to disabled first for at least 1 minor release.</li></ul> |
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+New feature gate unit tests will be added.
+
+##### Integration tests
+
+New feature gate integration tests will be added.
+
+##### e2e tests
+
+We will add one cluster level and one server level experimental features to the feature gate (without removing the original experimental flag), and add e2e tests to make sure the feature gate functions equivalently to their `--experimental-xxx` flags.
+
+We will also add downgrade/upgrade e2e tests to make sure the feature gate does not break the cluster in mixed version cluster.
+
+### Graduation Criteria
+
+#### Milestone 1
+
+* server level feature gate implemented.
+* cluster level feature gate implemented.
+
+#### Milestone 2
+
+* backport necessary proto changes to 3.5.
+* http endpoints added.
+* grpc endpoints added.
+* `etcdctl` command added.
+* equivalent of one cluster level and one server level experimental features added to the feature gate.
+* initial e2e tests completed.
+
+#### Milestone 3
+
+* e2e tests for upgrade & downgrade & mixed version scenarios completed.
+* robustness tests completed. 
+* documentation.
+
+### Upgrade / Downgrade Strategy
+
+The feature gate feature would available in 3.6+. 
+In 3.5, the necessary proto changes would be backported, but feature gate would not be available. 
```

Comment:
Why do we need to backport proto change to 3.5? We are just adding the feature of feature gate. We do not add any new feature which may change or break the proto.

For the impact on the upgrade/downgrade, we need to analyze each new feature case by case. But for this feature (feature-gate), it seems safe for now.

If we are planning to migrate some existing experimental feature, we can do it in 3.7.

Reply:
The plan is for the cluster level features, each member needs to know the enabled features in other members, and save that in a new member Attribute filed.

Reply:
Let's try not impact 3.5 in any way. We should keep the downgrade as simple & easy as possible.

FYI. I am thinking to simplify the existing downgrade workflow. But it can be discussed separately.

Reply:
Added the `Risks and Mitigations` section to explain this change should not really impact 3.5. Does it make sense @ahrtr ?

Reply:
No longer in scope.